dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/RelationAnalysis.java

```diff
@@ -1043,7 +1048,34 @@ public Knowledge visitVirtualLocation(Relation rel) {
                     }
                 }
             }
-            return new Knowledge(must, new HashSet<>(must));
+            return new Knowledge(must, enableMustSets ? new HashSet<>(must) : EMPTY_SET);
+        }
+
+        @Override
+        public Knowledge visitSyncWith(Relation rel) {
+            Set<Tuple> must = new HashSet<>();
+            List<Event> events = new ArrayList<>();
+            events.addAll(program.getThreadEvents(Load.class));
+            events.addAll(program.getThreadEvents(Store.class));
+            events.addAll(program.getThreadEvents(Fence.class));
+            events.addAll(program.getThreadEvents(RMWXchgBase.class));
+            events.removeIf(e -> e instanceof Init);
+            for (Event e1 : events) {
+                for (Event e2 : events) {
+                    Thread thread1 = e1.getThread();
+                    Thread thread2 = e2.getThread();
+                    if (thread1 == thread2 || thread1.hasSyncSet() || thread2.hasSyncSet()) {
```

Comment:
I think this should be `thread1 == thread2 || !thread1.hasSyncSet() || !thread2.hasSyncSet()`

That being said ... the code seems to be assuming whenever the test has `ssw 0 1;`, we add 1 to `thread1.SyncSet` and 0 to `thread2.SyncSet`. I think our parser guarantees this, but if the implementation relies on this, we need to have this properly documented

Reply:
Updated, and added comment